doc/realm.1
'\" t
.\"     Title: realm
.\"    Author: Stef Walter <stef@thewalter.net>
.\" Generator: DocBook XSL Stylesheets v1.76.1 <http://docbook.sf.net/>
.\"      Date: 09/21/2012
.\"    Manual: User Commands
.\"    Source: realmd
.\"  Language: English
.\"
.TH "REALM" "1" "" "realmd" "User Commands"
.\" -----------------------------------------------------------------
.\" * Define some portability stuff
.\" -----------------------------------------------------------------
.\" ~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
.\" http://bugs.debian.org/507673
.\" http://lists.gnu.org/archive/html/groff/2009-02/msg00013.html
.\" ~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
.ie \n(.g .ds Aq \(aq
.el       .ds Aq '
.\" -----------------------------------------------------------------
.\" * set default formatting
.\" -----------------------------------------------------------------
.\" disable hyphenation
.nh
.\" disable justification (adjust text to left margin only)
.ad l
.\" -----------------------------------------------------------------
.\" * MAIN CONTENT STARTS HERE *
.\" -----------------------------------------------------------------
.SH "NAME"
realm \- Manage enrollment in realms
.SH "SYNOPSIS"
.HP \w'\fBrealm\ discover\ \fR\fB\-v\fR\fB\ \fR\fB[realm\-name]\fR\ 'u
\fBrealm discover \fR\fB\-v\fR\fB \fR\fB[realm\-name]\fR
.HP \w'\fBrealm\ join\ \fR\fB\-v\fR\fB\ \fR\fB[\-U\ user]\fR\fB\ \fR\fB[realm\-name]\fR\ 'u
\fBrealm join \fR\fB\-v\fR\fB \fR\fB[\-U\ user]\fR\fB \fR\fB[realm\-name]\fR
.HP \w'\fBrealm\ leave\ \fR\fB\-v\fR\fB\ \fR\fB[\-U\ user]\fR\fB\ \fR\fB[realm\-name]\fR\ 'u
\fBrealm leave \fR\fB\-v\fR\fB \fR\fB[\-U\ user]\fR\fB \fR\fB[realm\-name]\fR
.HP \w'\fBrealm\ list\fR\ 'u
\fBrealm list\fR
.HP \w'\fBrealm\ permit\ \fR\fB[\-a]\fR\fB\ \fR\fB[\-R\ realm]\fR\fB\ \fR\fB{user}\fR\fB\ \&.\&.\&.\fR\ 'u
\fBrealm permit \fR\fB[\-a]\fR\fB \fR\fB[\-R\ realm]\fR\fB \fR\fB{user}\fR\fB \&.\&.\&.\fR
.HP \w'\fBrealm\ deny\ \fR\fB[\-a]\fR\fB\ \fR\fB[\-R\ realm]\fR\fB\ \fR\fB{user}\fR\fB\ \&.\&.\&.\fR\ 'u
\fBrealm deny \fR\fB[\-a]\fR\fB \fR\fB[\-R\ realm]\fR\fB \fR\fB{user}\fR\fB \&.\&.\&.\fR
.SH "DESCRIPTION"
.PP
\fBrealm\fR
is a command line tool that can be used to manage enrollment in kerberos realms, like Active Directory domains or IPA domains\&.
.SH "DISCOVER"
.PP
Discover a realm and its capabilities\&.
.sp
.if n \{\
.RS 4
.\}
.nf
$ realm discover
.fi
.if n \{\
.RE
.\}
.sp
.if n \{\
.RS 4
.\}
.nf
$ realm discover domain\&.example\&.com
.fi
.if n \{\
.RE
.\}
.PP
After discovering a realm, its name, type and capabilities are displayed\&.
.PP
If no domain is specified, then the domain assigned through DHCP is used as a default\&.
.PP
The following options can be used:
.PP
\fB\-\-client\-software=xxx\fR
.RS 4
Only discover realms for which we can use the given client software\&. Possible values include
\fIsssd\fR
or
\fIwinbind\fR\&.
.RE
.PP
\fB\-\-server\-software=xxx\fR
.RS 4
Only discover realms for run the given server software\&. Possible values include
\fIactive\-directory\fR
or
\fIfreeipa\fR\&.
.RE
.PP
\fB\-\-verbose, \-v\fR
.RS 4
Display verbose diagnostics while doing the discovery\&.
.RE
.SH "JOIN"
.PP
Configure the local machine for use with a realm\&.
.sp
.if n \{\
.RS 4
.\}
.nf
$ realm join domain\&.example\&.com
.fi
.if n \{\
.RE
.\}
.sp
.if n \{\
.RS 4
.\}
.nf
$ realm join \-user=admin \-\-computer\-ou=OU=Special domain\&.example\&.com
.fi
.if n \{\
.RE
.\}
.PP
The realm is first discovered, as we would with the
\fBdiscover\fR
command\&. If no domain is specified, then the domain assigned through DHCP is used as a default\&.
.PP
After a successful join, the computer will be in a state where it is able to resolve remote user and group names from the realm\&. For kerberos realms, a computer account and host keytab is created\&.
.PP
Joining arbitrary kerberos realms is not supported\&. The realm must have a supported mechanism for joining from a client machine, such as Active Directory or FreeIPA\&.
.PP
Unless a
[\-\-user]
is explicitly specified, then if possible, an automatic join is attempted first\&.
.PP
The following options can be used:
.PP
\fB\-\-client\-software=xxx\fR
.RS 4
Only join realms for which we can use the given client software\&. Possible values include
\fIsssd\fR
or
\fIwinbind\fR\&. Not all values are supported for all realms\&. By default the client software is automatically selected\&.
.RE
.PP
\fB\-\-computer\-ou=OU=xxx\fR
.RS 4
The distinguished name of an organizational unit to create the computer account\&. The exact format of the distinguished name depends on the client software and membership software\&. You can usually omit the root DSE portion of distinguished name\&.
.RE
.PP
\fB\-\-membership\-software=xxx\fR
.RS 4
The software to use when joining to the realm\&. Possible values include
\fIsamba\fR
or
\fIadcli\fR\&. Not all values are supported for all realms\&. By default the membership software is automatically selected\&.
.RE
.PP
\fB\-\-one\-time\-password=xxxx\fR
.RS 4
Perform the join using a one time password specified on the command line\&. This is not possible with all types of realms\&.
.RE
.PP
\fB\-\-server\-software=xxx\fR
.RS 4
Only join realms for run the given server software\&. Possible values include
\fIactive\-directory\fR
or
\fIfreeipa\fR\&.
.RE
.PP
\fB\-\-user\fR
.RS 4
The user name to be used to authenticate with when joining the machine to the realm\&. You will be prompted for a password\&.
.RE
.PP
\fB\-\-verbose, \-v\fR
.RS 4
Display verbose diagnostics while doing the discovery and join\&.
.RE
.SH "LEAVE"
.PP
Deconfigure the local machine for use with a realm\&.
.sp
.if n \{\
.RS 4
.\}
.nf
$ realm leave
.fi
.if n \{\
.RE
.\}
.sp
.if n \{\
.RS 4
.\}
.nf
$ realm leave domain\&.example\&.com
.fi
.if n \{\
.RE
.\}
.PP
If no realm name is specified, then the first configured realm will be used\&.
.PP
The following options can be used:
.PP
\fB\-\-user\fR
.RS 4
The user name to be used to authenticate with when leaving the realm\&. You will be prompted for a password\&.
.RE
.PP
\fB\-\-verbose, \-v\fR
.RS 4
Display verbose diagnostics while doing the leave operation\&.
.RE
.SH "LIST"
.PP
List all the discovered and configured realms\&.
.sp
.if n \{\
.RS 4
.\}
.nf
$ realm list
.fi
.if n \{\
.RE
.\}
.PP
The following options can be used:
.PP
\fB\-\-verbose, \-v\fR
.RS 4
Display verbose diagnostics while listing\&.
.RE
.SH "PERMIT"
.PP
Permit local login by users of the realm\&.
.sp
.if n \{\
.RS 4
.\}
.nf
$ realm permit \-a
.fi
.if n \{\
.RE
.\}
.sp
.if n \{\
.RS 4
.\}
.nf
$ realm permit DOMAIN\eUser
.fi
.if n \{\
.RE
.\}
.PP
If more than one realm is configured, then use the
\fB\-\-realm\fR
option to specify which realm to permit the users on\&.
.PP
The format of the user name can be seen by using the
\fBlist\fR
command\&.
.PP
The following options can be used:
.PP
\fB\-\-all, \-a\fR
.RS 4
Permit login by any valid user of the realm\&.
.RE
.PP
\fB\-\-realm, \-R\fR
.RS 4
Specify the name of the realm to permit users to log into\&.
.RE
.PP
\fB\-\-verbose, \-v\fR
.RS 4
Display verbose diagnostics while doing the operation\&.
.RE
.SH "DENY"
.PP
Deny local login by users of the realm\&.
.sp
.if n \{\
.RS 4
.\}
.nf
$ realm deny \-a
.fi
.if n \{\
.RE
.\}
.sp
.if n \{\
.RS 4
.\}
.nf
$ realm deny DOMAIN\eUser
.fi
.if n \{\
.RE
.\}
.PP
If more than one realm is configured, then use the
\fB\-\-realm\fR
option to specify which realm to deny the users\*(Aq login via\&.
.PP
The format of the user name can be seen by using the
\fBlist\fR
command\&.
.PP
The following options can be used:
.PP
\fB\-\-all, \-a\fR
.RS 4
Deny login by any validuser of the realm\&.
.RE
.PP
\fB\-\-realm, \-R\fR
.RS 4
Specify the name of the realm to deny users login to\&.
.RE
.PP
\fB\-\-verbose, \-v\fR
.RS 4
Display verbose diagnostics while doing the operation\&.
.RE


NEWS
0.8.1
 * Distribute debian settings file correctly

0.8
 * Cleaner, faster and robuster IPA discovery
 * Don't refret the PackageKit cache when installing
 * Add first translation, although gettext not hooked up yet
 * Pass passwords directly to the samba 'net' command
 * Better prompting for realm administrator/user password in realm tool
 * Remove custom debug code, just use glib
 * Support basic one-time-password and automatic joining
 * Ignore temporarily DNS failures when discovering
 * Add a generic kerberos provider, for discovering non AD/IPA realms
 * Add support for specifying a computer OU LDAP DN when joining
 * Don't require unstable glib
 * Fix timeout in the realm tool
 * Some build and testing fixes
 * Other bug fixes

0.7
 * Refactor the DBus interface so we can add other realm types later
 * Implement o.f.DBus.ObjectManager
 * Some compat fixes for samba4
 * Use better LoginFormats format
 * Add DBus interface documentation
 * Build fixes

0.6
 * Switch default AD client to sssd
 * Support building with automake 1.12.x
 * Cleaner DBus API which handles password based authentication
   required for samba3's net command.
 * Remove support for out of process providers
 * Lots of bug fixes

0.5
 * Implement login policy concept
 * Make the daemon exit-after-timeout work better
 * Implement an AD provider based on sssd
 * Fix accidental corruption of config files
 * New 'realm' command with various sub-commands
 * Display errors correctly in command line tools
 * Add discovery of IPA servers
 * Various bug fixes

0.4
 * Fix up license headers
 * Don't use SOA for MSDCS subdomain, use SRV record directly

0.3
 * Build and packaging fixes

0.2
 * Add possibility for cancellation to the DBus interface
 * Add provider Name and Version properties
 * Clearer diagnostics on discovery
 * Add realm-discover tool to identify a domain
 * Add proper DBus service files

0.1
 * Initial release


tools/realm.c
/* realmd -- Realm configuration service
 *
 * Copyright 2012 Red Hat Inc
 *
 * This program is free software: you can redistribute it and/or modify
 * it under the terms of the GNU Lesser General Public License as published
 * by the Free Software Foundation; either version 2 of the licence or (at
 * your option) any later version.
 *
 * See the included COPYING file for more information.
 *
 * Author: Stef Walter <stefw@gnome.org>
 */

#include "config.h"

#include "realm.h"
#include "realm-dbus-constants.h"

#include <glib.h>
#include <glib/gi18n.h>
#include <glib/gstdio.h>
#include <glib-object.h>

#include <krb5/krb5.h>

#include <errno.h>
#include <fcntl.h>
#include <locale.h>

struct {
	const char *name;
	int (* function) (int argc, char *argv[]);
	const char *usage;
	const char *description;
} realm_commands[] = {
	{ "discover", realm_discover, "realm discover -v [realm-name]", N_("Discover available realm") },
	{ "join", realm_join, "realm join -v [-U user] realm-name", N_("Enroll this machine in a realm") },
	{ "leave", realm_leave, "realm leave -v [-U user] [realm-name]", N_("Unenroll this machine from a realm") },
	{ "list", realm_list, "realm list", N_("List known realms") },
	{ "permit", realm_permit, "realm permit [-a] [-R realm] user ...", N_("Permit user logins") },
	{ "deny", realm_deny, "realm deny [-a] [-R realm] user ...", N_("Deny user logins") },
};

void
realm_print_error (const gchar *format,
                   ...)
{
	GString *message;
	va_list va;

	va_start (va, format);

	message = g_string_new ("");
	g_string_append_printf (message, "%s: ", g_get_prgname ());

	va_start (va, format);
	g_string_append_vprintf (message, format, va);
	va_end (va);

	g_printerr ("%s\n", message->str);
	g_string_free (message, TRUE);
}

void
realm_handle_error (GError *error,
                    const gchar *format,
                    ...)
{
	GString *message;
	va_list va;

	message = g_string_new ("");
	g_string_append_printf (message, "%s: ", g_get_prgname ());

	if (format) {
		va_start (va, format);
		g_string_append_vprintf (message, format, va);
		va_end (va);
	}

	if (error) {
		g_dbus_error_strip_remote_error (error);
		if (format)
			g_string_append (message, ": ");
		g_string_append (message, error->message);
		g_error_free (error);
	}

	g_printerr ("%s\n", message->str);
	g_string_free (message, TRUE);
}

GVariant *
realm_build_options (const gchar *first,
                     ...)
{
	const gchar *value;
	GPtrArray *opts;
	GVariant *options;
	GVariant *option;
	va_list va;

	va_start (va, first);

	opts = g_ptr_array_new ();
	while (first != NULL) {
		value = va_arg (va, const gchar *);
		if (value != NULL) {
			option = g_variant_new ("{sv}", first, g_variant_new_string (value));
			g_ptr_array_add (opts, option);
		}

		first = va_arg (va, const gchar *);
	}

	va_end (va);

	options = g_variant_new_array (G_VARIANT_TYPE ("{sv}"), (GVariant * const*)opts->pdata, opts->len);
	g_ptr_array_free (opts, TRUE);

	return options;
}

static void
propagate_krb5_error (GError **dest,
                      krb5_context context,
                      krb5_error_code code,
                      const gchar *format,
                      ...)
{
	GString *message;
	va_list va;

	message = g_string_new ("");

	if (format) {
		va_start (va, format);
		g_string_append_vprintf (message, format, va);
		va_end (va);
	}

	if (code != 0) {
		if (format)
			g_string_append (message, ": ");
		g_string_append (message, krb5_get_error_message (context, code));
	}

	g_set_error_literal (dest, g_quark_from_static_string ("krb5"),
	                     code, message->str);
	g_string_free (message, TRUE);
}

static GVariant *
read_file_into_variant (const gchar *filename)
{
	GVariant *variant;
	GError *error = NULL;
	gchar *contents;
	gsize length;

	g_file_get_contents (filename, &contents, &length, &error);
	if (error != NULL) {
		realm_handle_error (error, _("Couldn't read credential cache"));
		return NULL;
	}

	variant = g_variant_new_from_data (G_VARIANT_TYPE ("ay"),
	                                   contents, length,
	                                   TRUE, g_free, contents);

	return g_variant_ref_sink (variant);
}

GVariant *
realm_kinit_to_kerberos_cache (const gchar *name,
                               const gchar *realm,
                               const gchar *password,
                               GError **error)
{
	krb5_get_init_creds_opt *options = NULL;
	krb5_context context = NULL;
	krb5_principal principal = NULL;
	krb5_error_code code;
	int temp_fd;
	gchar *full_name = NULL;
	gchar *filename = NULL;
	krb5_ccache ccache = NULL;
	krb5_creds my_creds;
	GVariant *result = NULL;

	code = krb5_init_context (&context);
	if (code != 0) {
		propagate_krb5_error (error, NULL, code, _("Couldn't initialize kerberos"));
		goto cleanup;
	}

	if (strchr (name, '@') == NULL)
		name = full_name = g_strdup_printf ("%s@%s", name, realm);

	code = krb5_parse_name (context, name, &principal);
	if (code != 0) {
		propagate_krb5_error (error, context, code, _("Couldn't parse user name: %s"), name);
		goto cleanup;
	}

	code = krb5_get_init_creds_opt_alloc (context, &options);
	if (code != 0) {
		propagate_krb5_error (error, context, code, _("Couldn't setup kerberos options"));
		goto cleanup;
	}

	filename = g_build_filename (g_get_user_runtime_dir (), "realmd-krb5-cache.XXXXXX", NULL);
	temp_fd = g_mkstemp_full (filename, O_RDWR, S_IRUSR | S_IWUSR);
	if (temp_fd == -1) {
		int errn = errno;
		g_set_error (error, G_FILE_ERROR, g_file_error_from_errno (errn),
		             _("Couldn't create credential cache file: %s"), g_strerror (errn));
		goto cleanup;
	}
	close (temp_fd);

	code = krb5_cc_resolve (context, filename, &ccache);
	if (code != 0) {
		propagate_krb5_error (error, context, code, _("Couldn't resolve credential cache"));
		goto cleanup;
	}

	code = krb5_get_init_creds_opt_set_out_ccache (context, options, ccache);
	if (code != 0) {
		propagate_krb5_error (error, context, code, _("Couldn't setup credential cache"));
		goto cleanup;
	}

	code = krb5_get_init_creds_password (context, &my_creds, principal, (char *)password,
	                                     password ? NULL : krb5_prompter_posix,
	                                     0, 0, NULL, options);
	if (code == KRB5KDC_ERR_PREAUTH_FAILED) {
		propagate_krb5_error (error, context, code, _("Invalid password for %s"), name);
		goto cleanup;
	} else if (code != 0) {
		propagate_krb5_error (error, context, code, _("Couldn't authenticate as %s"), name);
		goto cleanup;
	}

	krb5_cc_close (context, ccache);
	ccache = NULL;

	result = read_file_into_variant (filename);
	krb5_free_cred_contents (context, &my_creds);

cleanup:
	g_free (full_name);

	if (filename) {
		g_unlink (filename);
		g_free (filename);
	}

	if (options)
		krb5_get_init_creds_opt_free (context, options);
	if (principal)
		krb5_free_principal (context, principal);
	if (ccache)
		krb5_cc_close (context, ccache);
	if (context)
		krb5_free_context (context);
	return result;
}

static int
usage (int code)
{
	gint i;

	for (i = 0; i < G_N_ELEMENTS (realm_commands); i++) {
		if (i > 0)
			g_printerr ("\n");
		g_printerr (" %s\n", realm_commands[i].usage);
		g_printerr ("   %s\n", realm_commands[i].description);
	}

	return code;
}

int
main (int argc,
      char *argv[])
{
	const gchar *command = NULL;
	gint i;

	setlocale (LC_ALL, "");

#ifdef ENABLE_NLS
	bindtextdomain (GETTEXT_PACKAGE, LOCALEDIR);
	bind_textdomain_codeset (GETTEXT_PACKAGE, "UTF-8");
	textdomain (GETTEXT_PACKAGE);
#endif

	g_type_init ();

	/* Find/remove the first non-flag argument: the command */
	for (i = 1; i < argc; i++) {
		if (command == NULL) {
			if (argv[i][0] != '-') {
				command = argv[i];
				argc--;
			}
		}
		if (command != NULL)
			argv[i] = argv[i + 1];
	}

	if (command == NULL)
		return usage (2);

	for (i = 0; i < G_N_ELEMENTS (realm_commands); i++) {
		if (g_str_equal (realm_commands[i].name, command))
			return (realm_commands[i].function) (argc, argv);
	}

	return usage(2);
}
